static/js/recent_senders.js

```diff
@@ -1,124 +1,185 @@
+import _ from "lodash";
+
 import {FoldDict} from "./fold_dict";
-import * as message_util from "./message_util";
+import * as message_store from "./message_store";
 
-// topic_senders[stream_id][topic_id][sender_id] = latest_message_id
-const topic_senders = new Map();
-// topic_senders[stream_id][sender_id] = latest_message_id
+class IdList {
+    ids = [];
+
+    // We cache the max message id to make sure that
+    // typeahead code is efficient.  We don't eagerly
+    // compute it, since it's plausible a spammy bot
+    // could cause us to process many messages at a time
+    // during fetching.
+    _cached_max_id = undefined;
+
+    add(id) {
+        this.ids.push(id);
+        this._cached_max_id = undefined;
+    }
+
+    remove(id) {
+        this.ids = _.pull(this.ids, id);
+        this._cached_max_id = undefined;
+    }
```

Comment:
I think we probably want to change `ids` to be a `Set` data structure rather than a list -- otherwise if you delete a topic that has 10K messages (E.g. from a bot that went crazy overnight), we'll be doing O(n) work for each ID O(n) times => O(n^2).

Reply:
Yeah, that's fair, I'll fix that.  (I also need to rebase after your other merge of `all_messages_data`.)